src/distances.jl

```diff
@@ -53,23 +66,83 @@ function fidelity(ψ::MPS, ρ::MPO)
   # TODO Check if trace is real
   #@assert imag(tr(ρ)) ≈ 0.0 atol=1e-5
   log_K = 2 * lognorm(ψ) + log(real(tr(ρ))) 
-  
   fidelity = exp(log_F̃ - log_K)
   return fidelity
 end
 
-# TODO: replace with:
-# X = L.X
-# log_F̃ = lognorm(X, ψ) # = loginner(X, ψ, X, ψ)
-# log_K = 2.0 * (lognorm(ψ) + lognorm(L))
-# return exp(log_F̃ - log_K)
-fidelity(ψ::MPS, L::LPDO{MPO}) = fidelity(ψ, MPO(L))
+fidelity(ρ::MPO, ψ::MPS) = fidelity(ψ, ρ)
 
-fidelity(ψ::MPS, L::LPDO{MPS}) = fidelity(ψ, L.X)
+"""
+Quantum state fidelity between an MPS wavefunction and a 
+LPDO density operator.
 
-fidelity(ρ::LPDO, ψ::MPS) = fidelity(ψ, ρ)
+F = ⟨ψ|ϱ|ψ⟩=|X†|ψ⟩|²
+"""
+function fidelity(Ψ::MPS, ϱ::LPDO{MPO})
+  proj = bra(ϱ) * Ψ
+  K = abs2(norm(Ψ)) * tr(ϱ)
+  return inner(proj,proj) / K
+end
+
+"""
+Quantum state fidelity between two MPO density matrices.
+
+F = (tr[√(√A B √A)])² 
+
+If `process = true`, get process fidelities:
+1. Quantum process fidelity between two MPO unitary operators.
+   F = |⟨⟨A|B⟩⟩|²
+2. Quantum process fidelity bewteen a MPO unitary and MPO Choi matrix
+   F = ⟨⟨A|B|A⟩⟩
+3. Quantum process fidelity bewteen two MPO Choi matrices
+   F = (Tr[√(√A B √A)])² 
+"""
+function fidelity(A::MPO, B::MPO; process::Bool = false, warnings::Bool = true)
+  # if quantum state fidelity:
+  if process
+    # check whether a MPO is Choi or not (from number of site indices)
+    # 1: unitary   -   unitary
+    (!_ischoi(A) && !_ischoi(B)) && return fidelity(_unitaryMPO_to_choiMPS(A), _unitaryMPO_to_choiMPS(B))
+    # 2: unitary   -   Choi MPO
+    (!_ischoi(A) && _ischoi(B)) && return _choifidelity(_unitaryMPO_to_choiMPS(A),B) 
+    # 2reverse: Choi MPO  -   unitary
+    (_ischoi(A) && !_ischoi(B)) && return _choifidelity(A,_unitaryMPO_to_choiMPS(B)) 
+  end
+  # quantum state/process fidelity bewtee two MPOs density matrices 
+  @disable_warn_order begin
```

Comment:
Could this be left up to the user to disable? It may be helpful for catching issues.

Reply:
Can you elaborate more? There is an explicit warning in the function being called (which can be deactivated), and the `@disable_warn_order` only propagates to a single function. I'm not sure what type of issues could be caught here. A better alternative would be to use the `warnings` argument to deactivate this as well, but I'm not sure how that would work with a macro, without having to completely replicate the code inside the `fidelity` for ITensors.

Reply:
Oh I see below...

Reply:
`@disable_warn_order` doesn't only propagate to a single function, it turns off the warning globally in all code wrapped in the macro.

Reply:
Right, but I did wrap only a single function, no?
```julia
  # quantum state/process fidelity bewtee two MPOs density matrices 
  @disable_warn_order begin
  return fidelity(prod(A), prod(B); warnings = warnings)
  end 
```

Reply:
I'm not saying you are using the macro wrong, I'm just arguing that we should allow users to control this externally and not turn it off internally in PastaQ.

Reply:
Ok,  I thought you missed that for some reason. I will open an issue on this.

Reply:
#183